Docs: Using TEAL debugger in algob (#419)
scale-it/algo-builder: docs/guide/teal-debugger.md

---
layout: splash
---

# About TEAL Debugger

Algorand provides the `tealdbg` command-line tool to launch an interactive session where the smart contract can be examined as the contract is being evaluated. The debugger supports both stateful and stateless smart contracts. You can debug individual transactions or atomic transfers that contain multiple, grouped transactions. The debugger supports setting the specific context for debugging purposes, including transactions, round number, latest timestamp, balance records, etc. The debugger runs either local programs or accepts HTTP connections from remote evaluators configured to run with a remote debugger hook.

# Using teal debugger with algob

> Setting up a `--dryrun-dump` of transaction data (via `goal` or SDK) could be a lengthy process, especially when using a transaction group. `Algob` provides an easy way to use debugger: by just supplying the transactions as an input to the `TealDbg` method (same transaction data that we supply to [executeTransaction](https://algobuilder.dev/api/algob/modules.html#executetransaction) to execute same transaction on network).

Reviewer: ```suggestion
Setting up a `--dryrun-dump` of transaction data (via `goal` or SDK) could be a lengthy process, especially when using a transaction group. `Algob` provides an easy way to use debugger: by simply supplying the transactions as an input to the `TealDbg` method (same transaction parameters that we supply to [executeTransaction](https://algobuilder.dev/api/algob/modules.html#executetransaction) to execute same transaction on network).
```

> ## Using Dryrun for Debugging a TEAL Program in an algob script

Reviewer 2: Let's use dry run instead of dryrun everywhere

> ## Using Dryrun for Debugging a TEAL Program in an algob script

Author: ok

Algob provides the functionality to do a test run of a TEAL smart contract. This option is useful to not only capture the transaction in the output file but the associated state of a smart contract. This allows testing of the TEAL logic in a dry run state which shows how the TEAL is processed and approved or rejected.

The dryrun response to this REST API includes disassembly, logic signature messages with PASS/REJECT, a signature trace, app call messages, and an app call trace.

NOTE: The dryrun REST API is only available on a node if it has been enabled in the node's configuration (`EnableDeveloperAPI` = true).
Example:
```js
// txnParams are the input transactions
const debugger = new Tealdbg(deployer, txnParam);
await debugger.dryRunResponse('dryrun.json');
```

> The DryRun Response will be dumped to `assets/dryrun.json`.

Reviewer: shouldn't this be in artifacts/cache ?

Author: no, I don't think it's an intermediate file, user will inspect the response after execution. Should this be in artifacts/cache?

Reviewer: yes I think so, because user doesn't create this. So it should be a part of artifacts

Author: ok, will update while adding support for stateful. Let's dump in `artifacts/dryrun/**`. But this means these files will be cleared by `algob clean`.

The smart contract used is [4-gold-asa.teal](https://github.com/scale-it/algo-builder/blob/master/examples/asa/assets/teal/4-gold-asa.teal) which ensures:
+ Transaction type is asset transfer and AssetAmount <= 1000.
+ Sender is `goldOwner` account.

First we need to deploy the contracts using `algob deploy`. We will debug the transaction present in [`scripts/transfer/gold-delegated-lsig.js`](https://github.com/scale-it/algo-builder/blob/master/examples/asa/scripts/transfer/gold-delegated-lsig.js).

### DryRun
Setting up transaction params (note that this is a passing scenario as amount = 500 <= 1000):
```js
// load signed lsig from checkpoint
const lsigGoldOwner = deployer.getDelegatedLsig('4-gold-asa.teal');
const txnParam = {
  type: types.TransactionType.TransferAsset,
  sign: types.SignType.LogicSignature,
  fromAccountAddr: goldOwner.addr,
  toAccountAddr: john.addr,
  amount: 500,
  assetID: 'gold',
  lsig: lsigGoldOwner,
  payFlags: { totalFee: 1000 }
};
```
After setting up the transaction, let's try to execute dryrun transaction:
```
const debug = new Tealdbg(deployer, txnParam);
await debug.dryRunResponse('dryrun-pass.json');
```

> Content of `assets/dryrun-pass.json` looks like (notice the logic-sig-message is PASS):

Reviewer: this should be artifacts right ?

---
layout: splash
---

> # About TEAL Debugger

Reviewer 2: ```suggestion
# Debugging TEAL
```

Reviewer 2: We use the top header (#) for the title. So there should be only one such header.

Author: Ok, thanks for info

Algorand provides the `tealdbg` command-line tool to launch an interactive session where the smart contract can be examined as the contract is being evaluated. The debugger supports both stateful and stateless smart contracts. You can debug individual transactions or atomic transfers that contain multiple, grouped transactions. The debugger supports setting the specific context for debugging purposes, including transactions, round number, latest timestamp, balance records, etc. The debugger runs either local programs or accepts HTTP connections from remote evaluators configured to run with a remote debugger hook.

> # Using teal debugger with algob

Reviewer 2: ```suggestion
## Using TEAL debugger with algob
```

Setting up a `--dryrun-dump` of transaction data (via `goal` or SDK) could be a lengthy process, especially when using a transaction group. `Algob` provides an easy way to use debugger: by simply supplying the transactions as an input to the `TealDbg` method (same transaction parameters that we supply to [executeTransaction](https://algobuilder.dev/api/algob/modules.html#executetransaction) to execute same transaction on network).

NOTE: You use the `TealDbg` method in an algob script, which can be run using `algob deploy`/`algob run` commands.

> ## Using Dryrun for Debugging a TEAL Program in an algob script

Reviewer 2: ```suggestion
### Using Dryrun for Debugging a TEAL Program in an algob script
```

Algob provides the functionality to do a test run of a TEAL smart contract. This option is useful to not only capture the transaction in the output file but the associated state of a smart contract. This allows testing of the TEAL logic in a dry run state which shows how the TEAL is processed and approved or rejected.

The dryrun response to this REST API includes disassembly, logic signature messages with PASS/REJECT, a signature trace, app call messages, and an app call trace.

NOTE: The dryrun REST API is only available on a node if it has been enabled in the node's configuration (`EnableDeveloperAPI` = true).
Example:
```js
// txnParams are the input transactions
const debugger = new Tealdbg(deployer, txnParam);
await debugger.dryRunResponse('dryrun.json');
```

The DryRun Response will be dumped to `assets/dryrun.json`.

> ## Starting a debugging session

Reviewer 2: ```suggestion
##$ Starting a debugging session
```

Author: I think this should be `###`

* `--group-index`: In case of a transaction group, group index should be passed to specify the current transaction in group in a debugger session.

**NOTE:** Passing `tealFile` is *optional*, but *recommended*. If not passed, debugger is run with the decompiled version of teal code. Supplying the program will allow debugging the original source and not the decompiled version.

```js
// txnParams are the input transactions
const debugger = new Tealdbg(deployer, txnParam);
await debugger.run({ tealFile: '4-gold-asa.teal' }); // 4-gold-asa.teal present in assets/**/
```

NOTE: To configure the chrome listener for a debugging session, read [Configure the Listener](https://github.com/algorand/go-algorand/blob/master/cmd/tealdbg/README.md#configure-the-listener).

## Example Walkthrough

In this section we will try to debug a stateless transaction in [`/examples/asa`](https://github.com/scale-it/algo-builder/tree/master/examples/asa).

The smart contract used is [4-gold-asa.teal](https://github.com/scale-it/algo-builder/blob/master/examples/asa/assets/teal/4-gold-asa.teal) which ensures:
+ Transaction type is asset transfer and AssetAmount <= 1000.
+ Sender is `goldOwner` account.

> First we need to deploy the contracts using `algob deploy`. We will debug the transaction present in [`scripts/transfer/gold-delegated-lsig.js`](https://github.com/scale-it/algo-builder/blob/master/examples/asa/scripts/transfer/gold-delegated-lsig.js).

Reviewer 2: ```suggestion
First we need to deploy the contracts using `algob deploy`. We will debug a transaction defined in [`scripts/transfer/gold-delegated-lsig.js`](https://github.com/scale-it/algo-builder/blob/master/examples/asa/scripts/transfer/gold-delegated-lsig.js).
```

### DryRun
Setting up transaction params (note that this is a passing scenario as amount = 500 <= 1000):
```js
// load signed lsig from checkpoint
const lsigGoldOwner = deployer.getDelegatedLsig('4-gold-asa.teal');
const txnParam = {
  type: types.TransactionType.TransferAsset,
  sign: types.SignType.LogicSignature,
  fromAccountAddr: goldOwner.addr,
  toAccountAddr: john.addr,
  amount: 500,
  assetID: 'gold',
  lsig: lsigGoldOwner,
  payFlags: { totalFee: 1000 }
};
```

> After setting up the transaction, let's try to execute dryrun transaction:

Reviewer 2: ```suggestion
After setting up the transaction, let's try to execute dry run by adding the following lines to the script:
```

```
const debug = new Tealdbg(deployer, txnParam);
await debug.dryRunResponse('dryrun-pass.json');
```

> Content of `assets/dryrun-pass.json` looks like (notice the logic-sig-message is PASS):

Reviewer 2: ```suggestion
It will create a `assets/dryrun-pass.json` file, which looks like (notice the logic-sig-message is PASS):
```

          "line": 3,
          "pc": 45,
          "stack": []
        },
        {
          "line": 4,
          "pc": 47,
          "stack": [
            {
              "bytes": "",
              "type": 2,
              "uint": 1
            }
          ]
        },
        ...
```

### Starting a debugging session

Let's try to launch a new debugger session with the same txnParam as in the above section:

```js
await debug.run({ tealFile: '4-gold-asa.teal' });
```

This will start a new debugger session. Console looks like:
```bash
2021/07/16 04:29:12 Using proto: https://github.com/algorandfoundation/specs/tree/d050b3cade6d5c664df8bd729bf219f179812595
2021/07/16 04:29:12 Run mode: logicsig
2021/07/16 04:29:12 ------------------------------------------------
2021/07/16 04:29:12 CDT debugger listening on: ws://127.0.0.1:9392/75c19568422ff120671707bc2682e7a3ae6861fe3bdac37571b860efd417e7d7
2021/07/16 04:29:12 Or open in Chrome:
2021/07/16 04:29:12 devtools://devtools/bundled/js_app.html?experiments=true&v8only=false&ws=127.0.0.1:9392/75c19568422ff120671707bc2682e7a3ae6861fe3bdac37571b860efd417e7d7
2021/07/16 04:29:12 ------------------------------------------------
```

> We now have a remote target set up in `chrome://inspect`

Reviewer 2: ```suggestion
Now, we have a remote target set up in `chrome://inspect`
```
